index.d.ts

```diff
@@ -0,0 +1,405 @@
+  'fantasy-land/chain': <B>(f: (A) => Maybe<B>) => Maybe<B>
+  'fantasy-land/alt': (other: Maybe<A>) => Maybe<A>
+  'fantasy-land/reduce': <B>(f: (B, A) => B, x: B) => B
+  'fantasy-land/traverse': <B>(typeRep: TypeRep, f: (A) => Applicative<B>) => Applicative<Maybe<B>>
+}
+
+export interface Either<A, B> {
+  'fantasy-land/equals': (other: Either<A, B>) => boolean
+  'fantasy-land/map': <C>(f: (B) => C) => Either<A, C>
+  'fantasy-land/bimap': <C, D>(f: (A) => C, g: (B) => D) => Either<C, D>
+  'fantasy-land/ap': <C>(other: Either<A, (B) => C>) => Either<A, C>
+  'fantasy-land/chain': <C>(f: (B) => Either<A, C>) => Either<A, C>
+  'fantasy-land/alt': (other: Either<A, B>) => Either<A, B>
+  'fantasy-land/reduce': <C>(f: (C, B) => C, x: C) => C
+  'fantasy-land/traverse': <C>(typeRep: TypeRep, f: (B) => Applicative<C>) => Applicative<Either<A, C>>
+}
+
+
+type ValidNumber = number;
+type FiniteNumber = number;
+type NonZeroFiniteNumber = number;
+type Integer = number;
+type NonNegativeInteger = number;
+
+
+interface TypeRep {}
+
+
+interface Setoid<A> {
+  'fantasy-land/equals': (other: Setoid<A>) => boolean
+}
+interface Ord<A> extends Setoid<A> {
+  'fantasy-land/lte': (other: Ord<A>) => boolean
+}
+interface Semigroupoid<A, B> {
+  'fantasy-land/compose': <C>(other: Semigroupoid<B, C>) => Semigroupoid<A, C>
+}
+interface Category<A> extends Semigroupoid<A, A> {
+  constructor: {
+    'fantasy-land/id': () => Category<A>
+  }
+}
+interface Semigroup<A> {
+  'fantasy-land/concat': (other: Semigroup<A>) => Semigroup<A>
+}
+interface Monoid<A> extends Semigroup<A> {
+  constructor: {
+    'fantasy-land/empty': () => Monoid<A>
+  }
+}
+interface Functor<A> {
+  'fantasy-land/map': <B>(f: (A) => B) => Functor<B>
+}
+interface Bifunctor<A, C> extends Functor<C> {
+  'fantasy-land/bimap': <B, D>(f: (A) => B, g: (C) => D) => Bifunctor<B, D>
+}
+interface Profunctor<B, C> extends Functor<C> {
+  'fantasy-land/promap': <A, D>(f: (A) => B, g: (C) => D) => Profunctor<A, D>
+}
+interface Apply<A> extends Functor<A> {
+  'fantasy-land/ap': <B>(other: Apply<(A) => B>) => Apply<B>
+}
+interface ApplicativeTypeRep<A> {
+  'fantasy-land/of': (value: A) => Applicative<A>
+}
+interface Applicative<A> extends Apply<A> {
+  constructor: ApplicativeTypeRep<A>
+}
+interface Chain<A> extends Apply<A> {
+  'fantasy-land/chain': <B>(f: (A) => Chain<B>) => Chain<B>
+}
+interface ChainRec<A> extends Chain<A> {
+  constructor: {
+    'fantasy-land/chainRec': <B, C>(f: ($0: (A) => C, $1: (B) => C, $2: A) => ChainRec<C>, x: A) => ChainRec<B>
+  }
+}
+interface Monad<A> extends Applicative<A>, Chain<A> {}
+interface Alt<A> extends Functor<A> {
+  'fantasy-land/alt': (other: Alt<A>) => Alt<A>
+}
+interface PlusTypeRep<A> {
+  'fantasy-land/zero': () => Plus<A>
+}
+interface Plus<A> extends Alt<A> {
+  constructor: PlusTypeRep<A>
+}
+interface Alternative<A> extends Applicative<A>, Plus<A> {
+  constructor: ApplicativeTypeRep<A> & PlusTypeRep<A>
+}
```

Comment:
I'm unhappy with this definition. I had hoped that

```typescript
interface Alternative<A> extends Applicative<A>, Plus<A> {}
```

would suffice, but the `constructor` property requirements are seen to conflict rather than combine.

Is there an elegant solution to this problem?